src/object/mod.rs

```diff
@@ -144,17 +147,47 @@ impl TopologyObject {
         unsafe { ffi::deref_str(&self.0.subtype) }
     }
 
+    /// Set the subtype string
+    ///
+    /// This exposes [`TopologyObject::set_subtype_unchecked()`] as a safe
+    /// method on operating systems which aren't known to facilitate mixing and
+    /// matching libc versions between an application and its dependencies.
+    #[allow(clippy::missing_errors_doc)]
+    #[cfg(all(feature = "hwloc-2_3_0", not(windows)))]
+    pub fn set_subtype(&mut self, subtype: &str) -> Result<(), NulError> {
+        // SAFETY: Underlying OS is assumed not to ergonomically encourage
+        //         unsafe multi-libc linkage
+        unsafe { self.set_subtype_unchecked(subtype) }
+    }
+
     /// Set the subtype string
     ///
     /// This is something you'll often want to do when creating Group or Misc
     /// objects in order to make them more descriptive.
     ///
+    /// # Safety
+    ///
+    /// This method is only safe to call if you can guarantee that your
+    /// application links against the same libc/CRT as hwloc.
+    ///
+    /// While linking against a different libc is something that is either
+    /// outright UB or strongly advised against on every OS, actually following
+    /// this sanity rule is unfortunately very hard on Windows, where usage of
+    /// multiple CRTs and pre-built DLLs is the norm, and there is no easy way
+    /// to tell which CRT version your pre-built hwloc DLL links against to
+    /// adjust your application build's CRT choice accordingly.
+    ///
+    /// Which is why hwlocality tries hard to accomodate violations of the rule.
+    /// And thus the safe version of this method, which assumes that you do
+    /// follow the rule, is not available on Windows.
+    ///
     /// # Errors
     ///
     /// - [`NulError`] if `subtype` contains NUL chars.
     #[cfg(feature = "hwloc-2_3_0")]
-    pub fn set_subtype(&mut self, subtype: &str) -> Result<(), NulError> {
-        self.0.subtype = LibcString::new(subtype)?.into_raw();
+    pub unsafe fn set_subtype_unchecked(&mut self, subtype: &str) -> Result<(), NulError> {
+        // SAFETY: Per input precondition
+        self.0.subtype = unsafe { LibcString::new(subtype)?.into_raw() };
         Ok(())
     }
```

Comment:
Safety fix brought by Windows CRT woes. Will get a more satisfactory fix once the next hwloc release introduces a proper way to set the subtype string to an hwloc-allocated string, not a user-allocated string. This will also trivially make the project rustix compatible as a bonus.